web/blog/tenzir-for-splunk-users/index.md

```diff
@@ -0,0 +1,316 @@
+## Why not SQL?
+
+Splunk was the first tool that provided an integrated solution from interactive
+data exploration to management-grade dashboards-all powered by dataflow
+pipelines. The success of Splunk is not only resulting from their first-mover
+advantage in the market, but also because their UX truly resonated with users:
+their target audience *liked* the user experience. It was *easy* to get things
+done.
+
+At Tenzir, we have a very clear target audience: security practitioners. They
+are not necessarily data engineers and fluent in SQL and low-level data tools,
+but rather identify as blue teamers, incident responders, threat hunters,
+detection engineers, threat intelligence analysts, and other domain experts. Our
+goal is cater to these folks, without requiring them to have deep understanding
+of relational algebra.
+
+The reason why we opted for a dataflow language is that it simplifies the
+reasoning: one step at a time. At least conceptually, because a smart system
+optimizes the execution of the users intent so that the pipeline runs
+efficiently. Any declarative language has that property, SQL in particular. But
+the difference to SQL is that a sequential dataflow makes it easy to simply
+*append* your next operation, as opposed to editing a query in the middle to
+inject a complicated clause somewhere. Instead of having to understand the
+entire SQL query, you just need to append an operation at the end that does the
+next thing. This dataflow pipeline style is becoming more an more popular. Most
+SIEMs have a language of their own, like Splunk. [Kusto][kusto] is another great
+example with a wide user base in security.
+
+[kusto]: https://learn.microsoft.com/en-us/azure/data-explorer/kusto/query/
+
+In fact, there's often an isomorphic SQL expression because a lot of the
+underlying engines internally map to the same Volcano-style model of execution.
+This gives rise to [transpiling dataflow languages to other execution
+platforms][splunk-transpiler]. Ultimately, our goal is that security
+practitioners do not have to think *any* of this and stay in their happy place,
+which means avoiding context switches to lower-level data primitives.
+
+[splunk-transpiler]: https://www.databricks.com/blog/2022/12/16/accelerating-siem-migrations-spl-pyspark-transpiler.html
+
+Now that we got the SQL topic out of the way, let's dive into some hands-on
+examples that illustrate the similarities and differences of SPL and TQL.
+
+## Examples
+
+Back in 2020, Eric Ooi wrote about [threat hunting with
+Zeek](https://www.ericooi.com/zeekurity-zen-part-iv-threat-hunting-with-zeek/),
+providing a set of Splunk queries that are corner stones for threat hunting.
+
+### Connections to destination port > 1024
+
+Splunk:
+
+```splunk-spl
+index=zeek sourcetype=zeek_conn id.resp_p > 1024
+| chart count over service by id.resp_p
+```
+
+Tenzir:
+
+```kusto
+export
+| where #schema == "zeek.conn" && id.resp_p > 1024
+| summarize count(.) by id.resp_p
+```
```

Comment:
Yep, I added `service`. 

In general, it's impossible to compare a chart with a data frame. That's why there's a note on `chart` a bit further down.

Reply:
[This blog post](https://www.splunk.com/en_us/blog/tips-and-tricks/search-commands-stats-chart-and-timechart.html) makes it seem like at least the underlying data is comparable.

Reply:
Thanks, this helps. I didn't realize that `chart` effectively does a `pivot_wider` on the second argument. Does your understanding match this?

```R
xs <- tribble(
   ~X, ~Y, ~Z,
  "a",  1, "foo",
  "b",  2, "bar",
  "a",  3, "foo",
  "b",  4, "bar",
  "c",  5, "foo"
)
```

```
❯ xs |>
   group_by(Z, X) |>
   summarize(N = sum(Y)) |>
   pivot_wider(names_from = X, values_from = N)
`summarise()` has grouped output by 'Z'. You can override using the `.groups`
argument.
# A tibble: 2 × 4
# Groups:   Z [2]
  Z         b     a     c
  <chr> <dbl> <dbl> <dbl>
1 bar       6    NA    NA
2 foo      NA     4     5
```

Since we don't have [`pivot_wider`](https://tidyr.tidyverse.org/reference/pivot_wider.html) yet, I think we need to rephrase this.

Reply:
Sorry, I'm not too familiar with R. But I think we are on the same page.

However, I think the `chart X over Y by Z` is kind of "isomorphic" to `summarize X by Y, Z`. Same data, different layout. So perhaps the comparison is still valid.

Reply:
In d2e0757ea1 I explained `chart` a bit more accurately.